File: bridge/src/hmac.js
import { createHash, createHmac, timingSafeEqual } from 'node:crypto';

export function canonicalJson(value) {
  if (value === null || typeof value !== 'object') {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalJson(item)).join(',')}]`;
  }
  return `{${Object.keys(value)
    .sort()
    .map((key) => `${JSON.stringify(key)}:${canonicalJson(value[key])}`)
    .join(',')}}`;
}

export function canonicalBodyHash(body) {
  return createHash('sha256').update(canonicalJson(body)).digest('hex');
}

export function pushSignaturePayload({ teamName, taskId, state, sequence, bodyHash }) {
  return [teamName, taskId, state, String(sequence), bodyHash].join('\n');
}

export function signPush({ secret, teamName, taskId, state, sequence, body }) {
  const bodyHash = canonicalBodyHash(body);
  return createHmac('sha256', secret)
    .update(pushSignaturePayload({ teamName, taskId, state, sequence, bodyHash }))
    .digest('hex');
}

export function verifyPushSignature({ expected, secret, teamName, taskId, state, sequence, body }) {
  const normalized = expected?.startsWith('sha256=') ? expected.slice('sha256='.length) : expected;
  if (!normalized || !/^[a-fA-F0-9]{64}$/.test(normalized)) return false;
  const actual = signPush({ secret, teamName, taskId, state, sequence, body });
  return timingSafeEqual(Buffer.from(actual, 'hex'), Buffer.from(normalized, 'hex'));
}

File: harness/bridge/hmac.py
from __future__ import annotations

import hashlib
import hmac
import json
import re
from typing import Any


def canonical_json(value: Any) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def canonical_body_hash(body: Any) -> str:
    return hashlib.sha256(canonical_json(body).encode("utf-8")).hexdigest()


def push_signature_payload(*, team_name: str, task_id: str, state: str, sequence: int | str, body_hash: str) -> str:
    return "\n".join([team_name, task_id, state, str(sequence), body_hash])


def sign_push(*, secret: str, team_name: str, task_id: str, state: str, sequence: int | str, body: Any) -> str:
    payload = push_signature_payload(
        team_name=team_name,
        task_id=task_id,
        state=state,
        sequence=sequence,
        body_hash=canonical_body_hash(body),
    )
    return hmac.new(secret.encode("utf-8"), payload.encode("utf-8"), hashlib.sha256).hexdigest()


def verify_push_signature(
    *,
    expected: str | None,
    secret: str,
    team_name: str,
    task_id: str,
    state: str,
    sequence: int | str,
    body: Any,
) -> bool:
    normalized = expected.removeprefix("sha256=") if expected else ""
    if not re.fullmatch(r"[a-fA-F0-9]{64}", normalized):
        return False
    actual = sign_push(secret=secret, team_name=team_name, task_id=task_id, state=state, sequence=sequence, body=body)
    return hmac.compare_digest(actual, normalized.lower())
